qiskit/primitives/utils.py

```diff
@@ -113,6 +113,30 @@ def final_measurement_mapping(circuit: QuantumCircuit) -> dict[int, int]:
     return mapping
 
 
+def _circuit_key(circuit: QuantumCircuit) -> tuple:
+    """Private key function for QuantumCircuit.
+
+    This is the workaround until :meth:`QuantumCircuit.__hash__` will be introduced.
+    If key collision is found, please add elements to avoid it.
+
+    Args:
+        circuit: Input quantum circuit.
+
+    Returns:
+        Key for directory.
+    """
+    return (
+        circuit.name,
+        circuit.num_qubits,
+        circuit.num_clbits,
+        circuit.num_parameters,
+        tuple(
+            (d.qubits, d.clbits, d.operation.name, tuple(d.operation.params)) for d in circuit.data
+        ),
+        None if circuit._op_start_times is None else tuple(circuit._op_start_times),
+    )
```

Comment:
sounds good to use `hash`

Reply:
But hash may collide, i.e. hash(-1) == hash(-2)

Reply:
I see. If we directly use a hash key, it may collide... The collision results in a wrong result as #8725. So, we have to use the tuple instead.

Reply:
Hash collisions are a possibility but very rare. No reasonable hash function is collision-free, but there even are cryptographically secure hashes. The largest the tuple, the less likely a collision.

Notice that these collisions can also happen in dictionary keys, and yet python rolls with it (which is a testament as to how unlikely collisions are). If you use a tuple as a dictionary key, internally python will call `hash()` anyway.

To be fair this is sometimes applied in tandem with `__eq__` so that equality can be checked after matching the hash (these are more advanced implementations of `Mapping`).

We could generate a bunch of random circuits and check for collisions under this paradigm to validate this.

Reply:
The intent here is to use it as a dictionary key, so you should use the tuple, not the hash of the tuple.  Let Python handle the hash collisions - you'll likely get it wrong if you do it yourself, and CPython's dict implementation is _extremely_ well battle-tested.  Effectively, hash collisions do happen frequently - Python's hash isn't cryptographic (hashes of <61-bit integers are just themselves), and hash-table implementations don't use the full range of the hash otherwise every table would require nearly the entire pointer space of a 64-bit system for the buckets alone.  CPython does it by just taking the x lowest bits (depending on how many buckets it's got allocated).

Since CPython non-empty dicts start with 8 buckets and resize when they exceed 2/3 load factor, with completely random hashes, you can fit 5 items in the collection before it expands.  If the buckets are assigned to completely randomly, you still expect to get a hash collision in 5-element dictionaries $1 - \frac{8! / (8-5)!}{8^5} \approx \frac45$ of the time.

Reply:
Python uses open addressing to resolve hash collision. This means that key is used to check the given objects are same or not. If key is hash value, we will not be able to avoid this collision. The example I gave is a rare case, but there are actually more collisions. ( It's natural because the hash will lose information.) Id collision exsists, it will be considered to be the same as the previous circuit, and the wrong value will be returned.

Reply:
That makes sense! Thanks for the detailed explanation 😄